execution/tvf/watermark_generator.go

```diff
@@ -118,6 +120,15 @@ func (s *WatermarkGeneratorStream) Next(ctx context.Context) (*execution.Record,
 	return srcRecord, nil
 }
 
-func (s *WatermarkGeneratorStream) Close() error {
-	return s.source.Close()
+func (s *WatermarkGeneratorStream) Close(ctx context.Context) error {
+	if err := s.source.Close(ctx); err != nil {
+		return errors.Wrap(err, "couldn't close underlying stream")
+	}
+
+	storage := storage.GetStateTransactionFromContext(ctx).GetUnderlyingStorage()
```

Comment:
Close isn't transactional, so there might not be a transaction here.

Reply:
Maybe the Close context should contain the Storage itself? Or Close should have a signature Close(ctx context.Context, storage storage.Storage) so that everybody gets it?

Reply:
I think expanding the signature by `storage.Storage` is a good and relatively easy to implement idea